apps/extension/src/background/approvals/service.ts

```diff
@@ -338,11 +307,54 @@ export class ApprovalsService {
     });
   };
 
-  private getPopupTabId = async (url: string): Promise<number | undefined> => {
-    const window = await this._launchApprovalWindow(url);
+  private getPopupTabId = (window: browser.Windows.Window): number => {
     const firstTab = window.tabs?.[0];
     const popupTabId = firstTab?.id;
 
+    if (!popupTabId) {
+      throw new Error("no popup tab ID");
+    }
+
+    if (popupTabId in this.resolverMap) {
+      throw new Error(`tab ID ${popupTabId} already exists in promise map`);
+    }
+
     return popupTabId;
   };
+
+  private launchApprovalPopup = async <T>(
```

Comment:
`launchApprovalPopup` now combines the same pattern we have in all functions: 
- Create the url
- Open the popup
- Return the resolver promise
- Plus: when the promise is resolved or rejected, it's also removed from the `resolverMap`